[5.5.x] rotate RPC credentials on cluster upgrade

a-palchikov opened the pull request:
## Description
<!--Required. Provide high-level overview of what the change is for.-->
* Rotate RPC credentials used for install/expand/leave operations on each cluster upgrade operation.
* Revise "system rotate-rpc-creds" to be able to use it on any affected `5.x` or `6.x` cluster and not necessarily based on `5.5.47`.

## Type of change
<!--Required. Keep only those that apply.-->

* Bug fix (non-breaking change which fixes an issue)
* New feature (non-breaking change which adds functionality)
* This change has a user-facing impact

## Linked tickets and other PRs
<!--Required. Keep only those that apply.-->

<!--This PR addresses the following issues.-->
* Updates https://github.com/gravitational/gravity/issues/1623

## TODOs
<!--Required. Keep only those that apply and check them off as they get completed.-->

- [x] Self-review the change
- [x] Perform manual testing
- [ ] Write documentation
- [ ] Address review feedback

## Implementation
<!--Optional. Add any relevant implementation details that might help the reviewers.-->

## Testing done
<!--Required. Explain what kind of testing these changes underwent.-->
Tested on `5.5.38` and `6.1.27`:
 * installed 3-node cluster
 * rotated the RPC certificates using the binary from this branch
 * removed a node
 * added the node back to cluster

The binary does not work on 7.x clusters due to changes in the manifest.

## Additional information
<!--Optional. Anything else that may be relevant.-->

a-palchikov committed:
Rotate RPC credentials used for install/expand/leave operations on each
upgrade operation.
Revise "system rotate-rpc-creds" to be able to use it on any affected
cluster and not necessarily based on 5.5.47-dev.x.

Updates https://github.com/gravitational/gravity/issues/1623

r0mant commented on tool/gravity/cli/clusterupdate.go:
```diff
@@ -60,6 +60,9 @@ func updateTrigger(
 	updatePackage string,
 	manual, noValidateVersion bool,
 ) error {
+	if err := updateRotateCertificates(localEnv); err != nil {
```
Just curious, any reason this is not done as a part of the "init" upgrade phase?

a-palchikov replied:
No reason - I did not think about it, will move it there.

r0mant reviewed (commented):
Regarding "testing done" - have you also tried performing an upgrade and verifying that credentials are rotated and that join/remove can be performed afterwards?

a-palchikov commented:
@r0mant yes, I tested upgrades `5.5.38` -> `5.5.47-dev.2`.

a-palchikov commented on tool/gravity/cli/clusterupdate.go:
```diff
@@ -96,6 +99,18 @@ func newClusterUpdater(
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	if _, err := rpc.UpsertCredentials(clusterPackages); err != nil {
```
I'll move that to the init phase and log there.

a-palchikov committed:
Address review comments

a-palchikov committed:
Address review comments

r0mant commented on lib/update/cluster/phases/init.go:
```diff
@@ -194,34 +194,64 @@ func (p *updatePhaseInit) Rollback(ctx context.Context) error {
+		return trace.Wrap(err)
 	}
+	defer rc.Close()
+	_, err = p.Packages.CreatePackage(rpcBackupPackage, rc, pack.WithLabels(env.RuntimeLabels))
```
Shouldn't this be "upsert"? Otherwise on the second upgrade this will fail because the package will already be there?

a-palchikov replied:
I've already made a change to assign it the corresponding operation label. Ideally, packages marked for the specific operation need to be purged post-operation. I'll make it into upsert though.

r0mant replied:
Yeah I think upsert is better anyway just so the phase is reentrable e.g. if reexecuted multiple times.

a-palchikov committed:
Create the RPC credentials backup package in cluster repository and
assign it the operation label so it gets automatically removed.
Also, upsert the package to be able to overwrite the contents and
labels.

r0mant committed:
Merge branch 'version/5.5.x' into dmitri/5.5.x/1623-rotate-rpc-creds

r0mant reviewed (commented):
@a-palchikov I have just realized that /init phase also executes on each server. Which means that credentials will be rotated multiple times and also each following package backup will override previous.

r0mant reviewed (approved):
Actually, I take my last comment back. It looks like it works as expected because there is a part of the init phase that is executed only on leader. I have also verified the following scenario: upgrade to this version, verify credentials have been rotated, verify remove, verify join. So this should be good to go. Going to merge once CI passes.
